crates/wasmtime/src/frame_info.rs

```diff
@@ -271,64 +277,57 @@ impl GlobalFrameInfo {
     /// is a wasm trap or not.
     pub(crate) fn is_wasm_pc(pc: usize) -> bool {
         let info = GLOBAL_INFO.lock().unwrap();
-        match func(pc, &info.ranges, |i| (i.start, &i.functions)) {
-            Some((_, info)) => info.instr_pos(pc).is_some(),
+
+        match info.ranges.range(pc..).next() {
+            Some((end, info)) => {
+                if pc < info.module.start || *end < pc {
+                    return false;
+                }
+
+                match info.module.func(pc) {
```

Comment:
The `is_wasm_pc` function is somewhat special in that it can primarily get invoked from a signal handler on Unix and Windows. Technically we're doing signal-unsafe things here by grabbing a lock as soon as we enter this function, but it's presumed fine in that we're pretty unlikely to segfault in this function itself. This reminds me, though, could you add some words to this function's documentation indicating that it's called from signal handlers and as a result the contents should be carefully written?

I was originally worried slightly that this call to `func` would involve assembling the `BTreeMap` which could involve allocations which could be tricky in a signal handler native-stack wise, but in thinking about it I'm not too worried about that.

Thinking a bit more about this, though, I'm not actually sure if we need the `BTreeMap` any more. Since `func_map()` is only ever used for function lookup, we can probably just do a binary search on an array with the pc? I presume that the list of functions is sorted in address order in some array in the compiled artifacts anyway. If we use that I think we can actually remove the need for that whole map entirely?

Reply:
I agree, we probably don't need a `BTreeMap` here.  I'll improve this further.

Reply:
I think we can guarantee that `finished_functions` are in sort-order (they're based on the ELF symbols which we are writing in-order) so we should be able to just utilize that for the search (based on the start vs. the end).